src/cryptenroll/cryptenroll-tpm2.c

```diff
@@ -58,11 +60,88 @@ static int search_policy_hash(
         return -ENOENT; /* Not found */
 }
 
+static int get_pin(char **ret_pin_str, int *ret_flags) {
+        char *pin_str = NULL;
+        int r = 0;
+        int flags = 0;
+        const char *e;
+        _cleanup_strv_free_erase_ char **pin = NULL;
+        _cleanup_strv_free_erase_ char **pin2 = NULL;
+
+        e = getenv("NEWPIN");
+        if (e) {
+                pin_str = strdup(e);
+                if (!pin_str)
+                        return log_oom();
+                flags |= TPM2_FLAGS_USE_PIN;
+
+                assert_se(unsetenv_erase("NEWPIN") >= 0);
+        } else {
+                for (int i = 5;; i--) {
+                        size_t pin_len;
+
+                        if (i <= 0) {
+                                return log_error_errno(
+                                        SYNTHETIC_ERRNO(ENOKEY), "Too many attempts, giving up.");
+                        }
+
+                        pin = strv_free_erase(pin);
+                        r = ask_password_auto(
+                                "Please enter TPM2 PIN:",
+                                "drive-harddisk",
+                                NULL,
+                                "tpm2-pin",
+                                "cryptsetup.tpm2-pin",
+                                USEC_INFINITY,
+                                0,
+                                &pin);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to ask for user pin: %m");
+                        assert(strv_length(pin) == 1);
+
+                        /* Enforce a PIN length of at least 4 characters (fewer than that provides no serious
+                         * protection) and 32 characters (maximum of TPM2 authValue). */
+                        pin_len = strlen(*pin);
+                        if (pin_len < 4 || pin_len > 32) {
```

Comment:
maybe encapsulate that in a simple helper `tpm2_valid_pin()` or so that does this check, and is used here and everywhere else where we need to superficially check if a password is suitable as TPM2 PIN